File: c++/51_N-Queens.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<string> > solveNQueens(int n) {
        vector<int> cols;
        vector<string> sol;
        vector<vector<string> > allSol;
        dfs(0, n, cols, sol, allSol);
        return allSol;
    }

    void dfs(int irow, int n, vector<int> cols, vector<string> sol, vector<vector<string> > &allSol) {
        if (irow == n) {
            allSol.push_back(sol);
            return;
        }
        for (int icol = 0; icol < n; icol++) {
            if (isValid(irow, icol, cols)) {
                string row(n, '.');
                row[icol] = 'Q';
                cols.push_back(icol);
                sol.push_back(row);
                dfs(irow + 1, n, cols, sol, allSol);
                cols.pop_back();
                sol.pop_back();
            }
        }

    }
    bool isValid(int irow, int icol, vector<int> cols) {
        for (int row = 0; row < cols.size(); row++) {
            if (icol == cols[row] or abs(irow - row) == abs(icol - cols[row]))
                return false;
        }
        return true;
    }
};


int main() {
	//Solution s;
    string s(5, '.');
}

File: python/51_N-Queens.py
#!/usr/bin/python
import unittest

class Solution(object):
    def solveNQueens(self, n):
        def isValid(irow, icol, cols):
            for row, col in enumerate(cols):
                # diff cols and diags...
                if icol == col or abs(irow - row) == abs(icol - col):
                    return False
                #if icol == col or abs(irow - icol) == abs(row - col): # [3,2] and [0, 1]
            return True
        def dfs(irow, cols, sol):
            if irow == n:
                allSol.append(sol)
                return
            for icol in range(n):
                if isValid(irow, icol, cols):
                    row = ['.'] * n
                    row[icol] = 'Q'
                    #sol.append("".join(row)) # important, this way leads to the change of final solution
                    dfs(irow + 1, cols + [icol], sol + ["".join(row)])
                    #sol.pop()

        # save coordinate of the columns
        allSol = []
        dfs(0, [], [])
        return allSol


def main():
    s = Solution()
    print s.solveNQueens(4)


if __name__ == "__main__":
    main()
